src/approaches/pg3_approach.py

```diff
@@ -46,14 +47,39 @@ def __init__(self, initial_predicates: Set[Predicate],
     def get_name(cls) -> str:
         return "pg3"
 
-    def _predict(self, state: State, atoms: Set[GroundAtom],
-                 goal: Set[GroundAtom]) -> _GroundNSRT:
-        del state  # unused
+    def predict(self, atoms: Set[GroundAtom],
+                goal: Set[GroundAtom]) -> _GroundNSRT:
+        """Predicts next GroundNSRT to be deployed based on the PG3 generated
+        policy."""
         ground_nsrt = utils.query_ldl(self._current_ldl, atoms, goal)
         if ground_nsrt is None:
             raise ApproachFailure("PG3 policy was not applicable!")
         return ground_nsrt
 
+    def solve(self, task: Task, timeout: int) -> Callable[[State], Action]:
+        _S: TypeAlias = FrozenSet[GroundAtom]
+        _A: TypeAlias = _GroundNSRT
+        skeleton = []
+        atoms_sequence = []
+        atoms = utils.abstract(task.init, self._initial_predicates)
+        atoms_sequence.append(atoms)
+        start_time = time.time()
+
+        while not task.goal.issubset(atoms):
+            ground_nsrt = self.predict(atoms, task.goal)
+            atoms = utils.apply_operator(ground_nsrt, atoms)
+            skeleton.append(ground_nsrt)
+            atoms_sequence.append(atoms)
+            if (time.time() - start_time) >= timeout:
```

Comment:
it doesn't matter much, but let's move this check up to the top of the loop, so that we terminate ASAP after time runs out